### PR TITLE
teeworlds: build on macOS

### PR DESCRIPTION
**Pull request by Luflosi**
###### Description of changes
Enable running teeworlds on macOS.

###### Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [x] aarch64-linux
  - [x] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [x] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [x] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [23.05 Release Notes (or backporting 22.11 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2305-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
  - [ ] (Release notes changes) Ran `nixos/doc/manual/md-to-db.sh` to update generated release notes
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

File: pkgs/games/teeworlds/default.nix [hunk 1]
```diff
@@ -20,21 +22,48 @@ stdenv.mkDerivation rec {
...
       --replace '#define DATA_DIR "data"' \
                 '#define DATA_DIR "${placeholder "out"}/share/teeworlds/data"'
+
+    # Quote nonsense is a workaround for https://github.com/NixOS/nix/issues/661
+    substituteInPlace 'other/bundle/client/Info.plist.in' \
+      --replace ${"'"}''${TARGET_CLIENT}' 'teeworlds' \
+      --replace ${"'"}''${PROJECT_VERSION}' '${version}'
+
+    # Apply https://github.com/teeworlds/teeworlds/commit/86d6687ef07f7f05457a7f67537b650656b13820
+    # Can't use fetchpatch or fetchpatch2 because of https://github.com/NixOS/nixpkgs/issues/32084
```
[the remark this answers is not in the document]
Comment (Luflosi) [reply pkgs/games/teeworlds/default.nix]: I guess I can put the patch into the list of `patches`, no need to use `patch -p1`. I don't want to use `fetchurl` to fetch the patch because
> Using fetchurl for dynamically generated patches (e.g. those from GitHub) is not safe since the file can change in the future

(from https://github.com/NixOS/nixpkgs/issues/32084#issuecomment-727223713)
so I included the patch as a file instead.

**Review by wegank — APPROVED** [review]

**Comment by Luflosi** [comment]
I removed the dependency on icoutils on Darwin.

**Review by wegank — APPROVED** [review]
Well, it works, despite the weird postInstall...

File: pkgs/games/teeworlds/default.nix [hunk 2]
```diff
@@ -15,26 +17,52 @@ stdenv.mkDerivation rec {
...
+    Carbon
+    Cocoa
   ];
 
-  postInstall = ''
+  postInstall = lib.optionalString stdenv.isLinux ''
     # Convert and install desktop icon
     mkdir -p $out/share/pixmaps
-    icotool --extract --index 1 --output $out/share/pixmaps/teeworlds.png $src/other/icons/teeworlds.ico
+    ${icoutils}/bin/icotool --extract --index 1 --output $out/share/pixmaps/teeworlds.png $src/other/icons/teeworlds.ico
```
Comment (SuperSandro2000) [inline pkgs/games/teeworlds/default.nix:53]: Why was this moved from nativeBuildInputs to here?

File: pkgs/games/teeworlds/default.nix [hunk 3]
```diff
@@ -51,7 +79,7 @@ stdenv.mkDerivation rec {
 
     homepage = "https://teeworlds.com/";
     license = "BSD-style, see `license.txt'";
-    maintainers = with lib.maintainers; [ astsmtl ];
-    platforms = lib.platforms.linux;
+    maintainers = with lib.maintainers; [ astsmtl Luflosi ];
+    platforms = lib.platforms.linux ++ lib.platforms.darwin;
```
Comment (SuperSandro2000) [inline pkgs/games/teeworlds/default.nix:83]:
```suggestion
    platforms = lib.platforms.unix;
```

File: pkgs/games/teeworlds/default.nix [hunk 4]
```diff
@@ -15,26 +17,52 @@ stdenv.mkDerivation rec {
     fetchSubmodules = true;
   };
 
+  patches = [
+    # Can't use fetchpatch or fetchpatch2 because of https://github.com/NixOS/nixpkgs/issues/32084
+    # Using fetchurl instead is also not a good idea, see https://github.com/NixOS/nixpkgs/issues/32084#issuecomment-727223713
+    ./rename-VERSION-to-VERSION.txt.patch
```
Comment (SuperSandro2000) [inline pkgs/games/teeworlds/default.nix:23]: We could also use some shell script.

File: pkgs/games/teeworlds/default.nix (hunk 2 above)
Comment (SuperSandro2000) [reply pkgs/games/teeworlds/default.nix:53]: for cross you would need to use buildPackages.icoutils but generally building nativeBuildInputs is preferred

File: pkgs/games/teeworlds/default.nix (hunk 4 above)
Comment (SuperSandro2000) [reply pkgs/games/teeworlds/default.nix:23]: Na, leave it as is.
